sdks/python/apache_beam/transforms/environments.py

```diff
@@ -0,0 +1,276 @@
+                container_image=self.container_image)
+            )
+  @Environment.register_urn(common_urns.environments.DOCKER.urn,
+                            beam_runner_api_pb2.DockerPayload)
+  def from_runner_api_parameter(payload, components, context):
+    return DockerEnvironment(container_image=payload.container_image)
+
+  @classmethod
+  def from_options(cls, options):
+    return cls(container_image=options.environment_config)
+
+
+class ProcessEnvironment(Environment):
+
+  env_type = 'PROCESS'
+  urn = common_urns.environments.PROCESS.urn
+
+  def __init__(self, command, os='', arch='', env=None):
+    self.command = command
+    self.os = os
+    self.arch = arch
+    self.env = env or {}
+
+  def to_runner_api_parameter(self, context):
+    return (common_urns.environments.PROCESS.urn,
+            beam_runner_api_pb2.ProcessPayload(
+                os=self.os,
+                arch=self.arch,
+                command=self.command,
+                env=self.env
+            ))
+
+  @Environment.register_urn(common_urns.environments.PROCESS.urn,
+                            beam_runner_api_pb2.ProcessPayload)
+  def from_runner_api_parameter(payload, components, context):
+    return ProcessEnvironment(command=payload.command, os=payload.os,
+                              arch=payload.arch, env=payload.env)
+
+  @classmethod
+  def from_options(cls, options):
+    config = json.loads(options.environment_config)
+    return cls(config.get('command'), os=config.get('os', ''),
+               arch=config.get('arch', ''), env=config.get('env', ''))
+
+
+class ExternalEnvironment(Environment):
+
+  env_type = ('EXTERNAL', 'LOOPBACK')
+  urn = common_urns.environments.EXTERNAL.urn
+
+  def __init__(self, url, params=None):
+    self.url = url
+    self.params = params
+
+  def to_runner_api_parameter(self, context):
+    return (common_urns.environments.EXTERNAL.urn,
+            beam_runner_api_pb2.ExternalPayload(
+                endpoint=endpoints_pb2.ApiServiceDescriptor(url=self.url),
+                params=self.params
+            ))
+
+  @Environment.register_urn(common_urns.environments.EXTERNAL.urn,
+                            beam_runner_api_pb2.ExternalPayload)
+  def from_runner_api_parameter(payload, components, context):
+    return ExternalEnvironment(payload.endpoint.url, params=payload.params)
+
+  @classmethod
+  def from_options(cls, options):
+    def looks_like_json(environment_config):
+      import re
+      return re.match(r'\s*\{.*\}\s*$', environment_config)
+
+    if looks_like_json(options.environment_config):
+      config = json.loads(options.environment_config)
+      url = config.get('url')
+      if not url:
+        raise ValueError('External environment endpoint must be set.')
+      params = config.get('params')
+    else:
+      url = options
+      params = None
+
+    return cls(url, params=params)
+
+
+class PortableEnvironment(Environment):
+
```

Comment:
I don't understand the purpose of this class.  The 'portable' urn did not exist before, so I think we may be doing something wrong by inventing it.  Why not use `ExternalEnvironment`?

Reply:
`ExternalEnvironment` requires `ExternalPayload` type, and in the `PortableRunner._create_environment` it was implying that if the `portable_options.environment_type` was passed in a urn that starts with `beam:env:` but not one of the urns in `DOCKER`, `PROCESS`, `EXTERNAL`, then it creates just an environment with the given urn and config as payload.

Reply:
my impression is that anything starting with `beam:env:` should be in python_urns.  let's go with that assumption.